### PR TITLE
Allow Hash as a possible config value

### PR DESCRIPTION
**Pull request by ojab**
It's handy to do something like
```ruby
v = {key: '"-----BEGIN RSA PRIVATE KEY-----…', key_password: 'qwerty'}
# and then
setting :openssl_key, constructor: ->(v) {
  OpenSSL::PKey.read(*v.values_at(:key, :key_password))
}
```
to have decrypted key in the settings instead of two strings, but dry-c complains that `nil` doesn't have `#update`. I suppose there is an expectation that if we're assigning a Hash — value is defined by `setting`, we shouldn't expect that.

### CODE REVIEW

**Comment by solnic** [comment]
I'm not sure what the actual problem is. Could you rephrase the description somehow? ie why is it trying to call `update` on `nil`?

**Comment by ojab** [comment]
```ruby
require 'dry-configurable'

class Xxx
  extend Dry::Configurable

  setting :xxx
end

Xxx.configure do |config|
  config.update(xxx: {})
end
```
leads to
```
…/dry-configurable-0.14.0/lib/dry/configurable/config.rb:62:in `block in update': undefined method `update' for nil:NilClass (NoMethodError)

            self[key].update(value)
                     ^^^^^^^
	from …/dry-configurable-0.14.0/lib/dry/configurable/config.rb:59:in `each'
```
because it checks if input value is a `Hash` and we're passing the hash, but `:xxx` is not a subconfig, but a plain value.

**Comment by ojab** [comment]
OTOH now we can overwrite `Config` with a plain value :/ I'll fix it a bit later today.

**Comment by ojab** [comment]
BTW, shouldn't `.is_a?(Hash)` be `.respond_to(:to_hash)` instead? It's a convention that `#to_hash` is explicit permission for an object to be represented as a `Hash`. 
```
irb(main):001:0> x = Object.new
=> #<Object:0x00007f250e9bf278>
irb(main):002:0> def x.to_hash; {x: 1, y: 2};end 
=> :to_hash
irb(main):003:0> {**x}
=> {:x=>1, :y=>2}
```

**Comment by ojab** [comment]
Added `ArgumentError` on assigning non-Hash to the setting.

**Comment by timriley** [comment]
This all makes sense and looks good to me! I'll update the is_a?(Hash) check to use `to_hash` in a follow-up PR, then will cut a release.

**Comment by timriley** [comment]
Released in 0.15.0: https://github.com/dry-rb/dry-configurable/releases/tag/v0.15.0
